### PR TITLE
Add phpstan

### PR DESCRIPTION
**Pull request by fezfez**
- Add phpstan
- Add strong type everywhere (needed by phpstan)
- Add unit test (99.13% coverage)
- DTO are now Rich Domain Model, witch prevent error when this lib will go with typed property (php 7.4)
- dead code removed
- `Cfonb120Reader` now return directly the statement list and can be used to read multiple file instead of being dedicated to read one file

This PR has BC, tagging a 2.0 is required.

### CODE REVIEW

**Comment by guillaume-sainthillier** [comment]
Hi, thanks for this PR. I'ill make an UPGRADE.md file to explain the BC changes.

As we dump to a new major version, I think we could change the `$operationDetails` getter to always return an array of `OperationDetail` instead of the first one. And remove the `getAllDetails` method so.

File: src/Banking/Statement.php [hunk 1]
```diff
@@ -23,6 +23,12 @@ class Statement
     /** @var Operation[] */
     private $operations = [];
 
+    public function __construct()
+    {
+        $this->oldBalance = null;
+        $this->newBalance = null;
```
Comment (guillaume-sainthillier) [inline src/Banking/Statement.php:29]: Could you move the `$operations = [];` in the constructor to be more consistent? As a general rule we will do all the property initialisations to the `__construct` method if you want to use the Rich Domain Model pattern now.

File: src/Banking/Statement.php [hunk 2]
```diff
@@ -35,6 +41,15 @@ public function getOldBalance(): ?Balance
         return $this->oldBalance;
     }
 
+    public function getOldBalanceOrThrowException(): Balance
```
Comment (guillaume-sainthillier) [inline src/Banking/Statement.php:44]: Do you really need this method ? I think we could do this in the original getter method and just add a `has<Old|New>Balance` instead.

WDYT?

File: src/Banking/Statement.php [hunk 3]
```diff
@@ -35,6 +41,15 @@ public function getOldBalance(): ?Balance
...
+    {
+        if ($this->oldBalance === null) {
+            throw new \RuntimeException('old balance is null');
```
Comment (guillaume-sainthillier) [inline src/Banking/Statement.php:47]: We could add a dedicated exception for this, WDYT?

**Review by guillaume-sainthillier — COMMENTED** [review]
Thanks for adding more tests btw!

File: src/Banking/Statement.php (hunk 3 above)
Comment (fezfez) [reply src/Banking/Statement.php]: i've added BalanceUnavailableException

File: src/Banking/Statement.php (hunk 2 above)
Comment (fezfez) [reply src/Banking/Statement.php]: done

File: src/Banking/Statement.php (hunk 1 above)
Comment (fezfez) [reply src/Banking/Statement.php:31]: good catch :)

**Comment by fezfez** [comment]
I have also added infection test, report say

```
255 mutations were generated:
     238 mutants were killed
       2 mutants were not covered by tests
      13 covered mutants were not detected
       2 errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 94%
         Mutation Code Coverage: 99%
         Covered Code MSI: 94%

```
it would be cool if we could kill those mutants before doing the 2.0.

**Comment by fezfez** [comment]
Some mutant cant actually be killed by design (AbstractCfonbParser::parseAmount, AbstractCfonbParser::parseDate, AbstractCfonbParser::parseLine)

**Comment by guillaume-sainthillier** [comment]
Hello Stephane, thank you for your commits, it's super clean.
I don't know what is mutant, but I agree if you want to kill them 👍

**Comment by fezfez** [comment]
infection testing is a way to check if tests have good assertion.
it re run the test with a modified AST, if tests did not fail with this change, a mutant has "escape".
A killed mutant mean that a test detect it (fail)

**Pull request merged**